Add a fast path for shaping ASCII text

mbrubeck opened the pull request:
On both my Linux laptop and iMac, this is about twice as fast as Harfbuzz text shaping on https://en.wikipedia.org/wiki/Barack_Obama.

I haven't tested this on any high-DPI (retina) displays, and I'm not 100% certain that the font unit scaling is correct there.

Depends on servo/core-text-rs#50.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11273)

<!-- Reviewable:end -->

mbrubeck committed:
Add a fast path for shaping ASCII text

mbrubeck committed:
Simplify FontTableMethods::with_buffer

mbrubeck committed:
Move hb_tag! macro and KERN into font module

mbrubeck committed:
Allow creation of unboxed FontTables

mbrubeck committed:
Make FontHandle fields private

mbrubeck committed:
Simplify FontHandle::new_from_template

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify gfx code, but no tests are modified. Please consider adding a test!

mbrubeck commented:
r? @pcwalton

pcwalton commented on components/gfx/platform/macos/font.rs:
```diff
@@ -155,12 +245,38 @@ impl FontHandleMethods for FontHandle {
+                let value = BigEndian::read_i16(&pairs[i * KERN_PAIR_LEN + 4..]);
+                trace!("Kerning for {}/{}: {:?}", first_glyph, second_glyph, value);
+                return value as f64 * table.scale;
+            }
```
Can this use `Vec::binary_search()`?

Even if not, I still think I'd rather have this be a separate function.

pcwalton commented on components/gfx/text/shaping/harfbuzz.rs:
```diff
@@ -207,7 +201,45 @@ impl ShaperMethods for Shaper {
+            !options.flags.contains(RTL_FLAG) &&
+            font.handle.can_do_fast_kerning() &&
+            text.is_ascii()
+        {
```
Can you factor this out into a separate method?

highfive commented:
New code was committed to pull request.

pcwalton commented on components/gfx/platform/freetype/font.rs:
```diff
@@ -73,22 +75,16 @@ impl FontHandleMethods for FontHandle {
+                  handle: fctx.clone(),
+                  can_do_fast_kerning: false,
+            };
+            handle.can_do_fast_kerning = handle.has_table(KERN) || !handle.has_table(GPOS);
```
Hmm. I would expect `handle.has_table(KERN) && !handle.has_table(GPOS)`. Am I missing something?

mbrubeck replied:
Switched to `has_table(KERN) && !has_table(GPOS)` for now.

highfive commented:
New code was committed to pull request.

mbrubeck commented:
Addressed review comments.

@pcwalton:

> Hmm. I would expect `handle.has_table(KERN) && !handle.has_table(GPOS)`. Am I missing something?

I assumed that if a font has both 'kern' and 'GPOS', then falling back to the 'kern' table would be sufficient for shaping ASCII-only strings.  It would be more conservative to use the slow path in this case, however.  This affects some commonly-used fonts like ArialMT and TimesNewRoman on Mac OS, and the DejaVu fonts on Linux.

Even if we implement partial GPOS support in the fast path, we'll need to make a similar assumption that ASCII characters aren't affected by any unsupported GPOS features (or, somehow scan the whole table to verify this).

We could also make this depend on the `text-rendering` CSS property…

pcwalton commented:
@mbrubeck I was actually assuming we'd scan the table to verify this. I looked at the tables in FontForge and they seemed to pass the test from first glance.

But this can be a followup.

highfive commented:
New code was committed to pull request.

highfive commented:
New code was committed to pull request.

mbrubeck commented:
Changed to fall back to the slow path for now if the font has `GPOS` and/or `GSUB` tables.

We'll need to implement both `GPOS` pair positioning and `GSUB` ligation to get the fast path working for most Latin text.

highfive commented:
New code was committed to pull request.

mbrubeck commented:
Whoops, messed up some tidy fixes. Will push a proper fix when I'm back at my computer.

highfive commented:
New code was committed to pull request.

pcwalton commented on components/gfx/font.rs:
```diff
@@ -24,6 +24,14 @@ use unicode_script::Script;
+    );
+}
+
+pub const KERN: u32 = ot_tag!('k', 'e', 'r', 'n');
```
Wish we had `const fn` :)

pcwalton commented on components/gfx/platform/freetype/font.rs:
```diff
@@ -74,11 +76,16 @@ impl FontHandleMethods for FontHandle {
                   handle: fctx.clone(),
+                  can_do_fast_shaping: false,
             };
+            // TODO: Implement basic support for GPOS and GSUB.
```
File an issue and link it here

mbrubeck replied:
Done.

highfive commented:
New code was committed to pull request.

highfive commented:
New code was committed to pull request.

pcwalton commented on components/gfx/platform/macos/font.rs:
```diff
@@ -61,8 +64,105 @@ impl FontTableMethods for FontTable {
+            }
+            let num_subtables = BigEndian::read_u16(&table[2..]);
+            let mut start = 4;
+            for _ in 0..num_subtables {
```
Kinda weird that the spec doesn't say what the subtable version number is supposed to be. Maybe add a TODO to maybe check this in the future?

mbrubeck replied:
Done.

pcwalton commented:
Looks good, just had a couple of nits!

highfive commented:
New code was committed to pull request.

mbrubeck commented:
@bors-servo r=pcwalton

* Fixed nits. Here we go…

bors-servo commented:
:pushpin: Commit 5991afa has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 5991afafa412cf1b36bf1d375fc8304328042496 -->

bors-servo commented:
:hourglass: Testing commit 5991afa with merge c0c70ef...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1023), [arm32](http://build.servo.org/builders/arm32/builds/1094), [arm64](http://build.servo.org/builders/arm64/builds/1077), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1094), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1150), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1124), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1107), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1150), [windows](http://build.servo.org/builders/windows/builds/800)

khaledhosny commented:
I think this is going backwards, Geko does not have a fast path and Blink is killing its fast path as well because it is proven to be a bad idea on the long run. HarfBuzz should be pretty fast when no OpenType layout is involved anyway, so if the code is slow then this is probably an issue in the Servo side.

CC @behdad.

pcwalton commented:
The fast path in WebKit doesn't do kerning at all from what I can tell.

By design, our fast path only kicks in when we detect that there will be no quality difference between HarfBuzz and what we do. It's been shown to result in a 2x speedup, and I don't see anything wrong that we're doing in Servo that would cause a slowdown.

khaledhosny commented:
I don’t know Rust that well, but the code seems to do the fast path when the font has a `kern` table and no `GPOS` or `GSUB`, but this is probably to match any AAT font, so on Mac OS X you won’t get any layout features for such fonts (not even kerning since will unlikely be a version 0 `kern` table). This would also match any Graphite font having a `kern` table (though I don’t know if it is common to have such a table in Graphite fonts).

Also if HarfBuzz path in Servo is slow, I think making it faster benefits everyone not only user of ASCII text and fonts with no OpenType layout table. Both Gecko and Blink cache the shaping result in various way to speed things up.

mbrubeck commented:
This is still an experiment; it remains to be seen whether it will still be faster than HarfBuzz on typical fonts once we add GPOS and GSUB support. Perhaps it won't be. If we do find any significant wins, I'll definitely see if they they can be shared with HarfBuzz.

Part of the speedup here may be from better inlining. (We might also be able to get this from Harfbuzz by compiling it with Clang and using cross-language LTO, which is also something we plan to pursue.) Part of it is also from writing data directly into Servo's internal data structures without allocating any intermediate buffers.  And having a separate path for ASCII inside of Servo might be a win whether or not we it is using HarfBuzz.

> this is probably to match any AAT font, so on Mac OS X you won’t get any layout features for such fonts

Yeah, we ought to disable the fast path if any ASCII character is covered by a table that it doesn't support.  I'm working on this.

behdad commented:
This is so not what we expect HarfBuzz clients to be doing...  My 0.02USD.
